src/theory/symmetry_detect.cpp

```diff
@@ -0,0 +1,525 @@
+
+        if (sub_expr1 == sub_expr2)
+        {
+          Trace("sym-dt") << "[sym-dt] Merge variable " << fst_var
+                          << " with variable " << snd_var << endl;
+          merged.insert(expr_it2->first);
+          part.push_back(snd_var);
+          p.d_var_to_subvar[snd_var] = expr_it1->first;
+        }
+      }
+      p.d_subvar_to_vars[expr_it1->first] = part;
+      Trace("sym-dt") << "[sym-dt] Add a new partition after merging: " << endl;
+      printPartition(p);
+      partitions.push_back(p);
+    }
+  }
+}
+
+void SymmetryDetect::processSingletonPartitions(
+    vector<Partition>& partitions, unordered_set<Node, NodeHashFunction>& vars)
+{
+  Trace("sym-dt") << "[sym-dt] Start post-processing partitions with size = "
+                  << partitions.size() << endl;
+
+  // Mapping between the substitution variable to the actual variable
+  map<Node, Node> subvar_to_var;
+  // Mapping between the substitution variable to the actual expression
+  map<Node, Node> subvar_to_expr;
+  // Partitions that have 2 or more variables
+  vector<Partition> new_partitions;
+  vector<Partition>::iterator part_it = partitions.begin();
+
+  // Collect singleton partitions: subvar_to_expr, subvar_to_var, and variables
+  while (part_it != partitions.end())
+  {
+    if (!(*part_it).d_var_to_subvar.empty()
+        && (*part_it).d_var_to_subvar.size() == 1)
+    {
+      vars.insert(((*part_it).d_var_to_subvar.begin())->first);
+      subvar_to_expr[((*part_it).d_var_to_subvar.begin())->second] =
+          (*part_it).d_term;
+      subvar_to_var[((*part_it).d_var_to_subvar.begin())->second] =
+          ((*part_it).d_var_to_subvar.begin())->first;
+    }
+    else if (!(*part_it).d_var_to_subvar.empty()
+             && (*part_it).d_var_to_subvar.size() >= 2)
+    {
+      map<Node, Node>::iterator var_to_subvar_it =
+          (*part_it).d_var_to_subvar.begin();
+
+      while (var_to_subvar_it != (*part_it).d_var_to_subvar.end())
+      {
+        vars.insert(var_to_subvar_it->first);
+        ++var_to_subvar_it;
+      }
+      // Only add partitions that have more than 1 variable
+      new_partitions.push_back(*part_it);
+    }
+    ++part_it;
+  }
+
+  // Save all partitions that have more than 1 variable
+  partitions = new_partitions;
+
+  // Do matches on singleton terms
+  if (!subvar_to_var.empty())
+  {
+    matches(partitions, subvar_to_var, subvar_to_expr);
+  }
+}
+
+void SymmetryDetect::collectChildren(Node node, vector<Node>& children)
+{
+  if (theory::quantifiers::TermUtil::isAssoc(node.getKind()))
+  {
+    for(const Node& n : node)
+    {
+      collectChildren(n, node.getKind(), children);
+    }
+  }
+  else
+  {
+    for(const Node& n : node)
+    {
+      children.push_back(n);
+    }
+  }
+}
+
+void SymmetryDetect::collectChildren(Node node, Kind k, vector<Node>& children)
+{
+  if (node.getKind() == k)
+  {
+    for(const Node& n : node)
+    {
+      collectChildren(n, k, children);
```

Comment:
Can we make this function non-recursive?

Reply:
This is to recursively to traverse a node and collect its children if they are under the same operator. I don't see any easy way to change it to non-recursive.